drafts/bitcartel-payment-disclosure/draft.rst

```diff
@@ -0,0 +1,404 @@
+- JoinSplitSig private key used to sign the JoinSplit transaction ``joinSplitPrivKey`` [#protocol]_ §4.4 Sending Notes
+
+When creating a transaction with one or more Sapling outputs, for each such output, a data structure is created to record:
+
+- transaction id
+- list of Output description indices [0..outlen-1]:
+- corresponding ocks
+- list of Spend description indices [0..inlen-1]
+
+- corresponding Spend descriptions:
+  - any (d, pk_d) s.t. pk_d = [ivk] GH(d) -> ivk -> (ak, nk)
+  - a ZIP 304 signature:
+    - nf
+    - (rk is implied)
+    - zkproof
+    - spendAuthSig
+
+What the prover needs:
+  have ask, α
+  pick fresh nsk
+  \alpha from original tx
+  derive (ak, nk) from (ask, nsk)
+  derive ivk = CRH^ivk(ak, nk)
+  pick a fresh d
+  derive g_d = GH(d), pk_d = [ivk] g_d
+  create a proof like ZIP 304:
+
+  - Let :math:`cm = \mathsf{NoteCommit}^\mathsf{Sapling}_0(\mathsf{repr}_\mathbb{J}(g_d), \mathsf{repr}_\mathbb{J}(pk_d), 1)`.
+
+  - Let :math:`rt` be the root of a Merkle tree with depth
+    :math:`\mathsf{MerkleDepth}^\mathsf{Sapling}` and hashing function
+    :math:`\mathsf{MerkleCRH}^\mathsf{Sapling}`, containing :math:`cm` at position 0, and
+    :math:`\mathsf{Uncommitted}^\mathsf{Sapling}` at all other positions.
+
+  - Let :math:`path` be the Merkle path from position 0 to :math:`rt`. [#merkle-path]_
+
+  - Let :math:`cv = \mathsf{ValueCommit}_0(1)`.
+
+    - This is a constant and may be pre-computed.
+
+  - Let :math:`nf = \mathsf{PRF}^\mathsf{nfSapling}_{\mathsf{repr}_\mathbb{J}(nk)}(\mathsf{repr}_\mathbb{J}(\mathsf{MixingPedersenHash}(cm, 0)))`.
+
+  - Let :math:`zkproof` be a Sapling spend proof with primary input :math:`(rt, cv, nf, rk)`
+    and auxiliary input :math:`(path, 0, g_d, pk_d, 1, 0, cm, 0, α, ak, nsk)`.
+    [#spend-statement]_
+
+  - Let :math:`rsk = \mathsf{SpendAuthSig.RandomizePrivate}(α, ask)`.
+
+  - Let :math:`coinType` be the 4-byte little-endian encoding of the coin type in its index
+    form, not its hardened form (i.e. 133 for mainnet Zcash).
+
+  - Let :math:`digest = \mathsf{BLAKE2b}\text{-}\mathsf{256}(\texttt{"ZIP304Signed"}\,||\,coinType, zkproof\,||\,msg)`.
+
+  - Let :math:`spendAuthSig = \mathsf{SpendAuthSig.Sign}(rsk, digest)`.
+
+- Return :math:`(d, pk_d, nf, zkproof, spendAuthSig)`.
+
+
+
+The payment disclosure data should be persisted to disk or a database so it can be retrieved later.
+
+When persisting, third party applications should expect serialization of payment disclosure to follow Zcash and upstream convention.
+
+::
+
+    ADD_SERIALIZE_METHODS;
+    ...
+    
+    READWRITE(marker);
+    READWRITE(version);
+    READWRITE(esk);
+    READWRITE(txid);
+    READWRITE(js);
+    READWRITE(n);
+    READWRITE(zaddr);
+    READWRITE(message);
+
+A new RPC call will be introduced to allow the sender to retrieve the payment disclosure data for a given shielded output index.
+
+``z_getpaymentdisclosure txid joinsplit_index output_index [message]``
```

Comment:
The `joinsplit_index output_index` here isn't sufficiently expressive. What about transparent, Sapling, and Orchard outputs?

Also, the sender may not need to reveal all inputs to show that sufficient value has been transferred. Maybe include an optional list of input indices to reveal in the disclosure.

Do Orchard actions have a similar issue to JoinSplits, where revealing part of it could also reveal the input, even if it's not one the sender wants to reveal?